x-pack/legacy/plugins/ml/public/util/custom_url_utils.ts

```diff
@@ -83,67 +91,86 @@ export function openCustomUrlWindow(fullUrl, urlConfig) {
 
 // Returns whether the url_value of the supplied config is for
 // a Kibana Discover or Dashboard page running on the same server as this ML plugin.
-function isKibanaUrl(urlConfig) {
+function isKibanaUrl(urlConfig: UrlConfig) {
   const urlValue = urlConfig.url_value;
   return urlValue.startsWith('kibana#/discover') || urlValue.startsWith('kibana#/dashboard');
 }
 
+/**
+ * Escape any double quotes in the value for correct use in KQL.
+ */
+function escapeForKQL(value: string): string {
+  return value.replace(/\"/g, '\\"');
+}
+
 // Builds a Kibana dashboard or Discover URL from the supplied config, with any
 // dollar delimited tokens substituted from the supplied anomaly record.
-function buildKibanaUrl(urlConfig, record) {
+function buildKibanaUrl(urlConfig: UrlConfig, record: AnomalyRecordSource) {
   const urlValue = urlConfig.url_value;
 
-  return String(urlValue).replace((/\$([^?&$\'"]{1,40})\$/g), (match, name) => {
-
-    // Use lodash get to allow nested JSON fields to be retrieved.
-    let tokenValue = _.get(record, name, null);
-
-    // If the token is an influencer, then the value in the record will be an array.
-    // For now just support passing the first influencer from the array.
-    // TODO - support passing multiple influencer values.
-    if (Array.isArray(tokenValue)) {
-      tokenValue = tokenValue[0];
-    }
-
-    if (tokenValue !== null && !(name === 'earliest' || name === 'latest')) {
-
-      if (urlValue.includes('language:lucene') === true) {
-        // Escape reserved characters if the query language is lucene (default was switched to KQL in 7.1).
-        tokenValue = `${escapeForElasticsearchQuery(tokenValue)}`;
-      } else {
-        // Escape any double quotes in the value for correct use in KQL.
-        tokenValue = tokenValue.replace(/\"/g, '\\"');
-      }
-
-      // Kibana URLs used rison encoding, so escape with ! any ! or ' characters
-      tokenValue = tokenValue.replace(/[!']/g, '!$&');
-
-      // URI encode in case of special characters in the value.
-      tokenValue = encodeURIComponent(tokenValue);
-    }
-
-    // If property not found string is not replaced.
-    return (tokenValue !== null) ? tokenValue : match;
-  });
+  const queryLanguageEscapeCallback = urlValue.includes('language:lucene')
+    ? escapeForElasticsearchQuery
+    : escapeForKQL;
+
+  // Compose callback for characters escaping and encoding.
+  const getResultTokenValue: (v: string) => string = _.flow(
```

Comment:
Nit: Maybe it's worth moving the `(v: string) => string` spec to its own type, I thought like now the line is a bit hard to read, at first glance I thought the `=>` was part of a factory function and not the type :)